acq4/devices/Sonicator.py

```diff
@@ -0,0 +1,210 @@
+from threading import RLock
+
+import numpy as np
+
+import pyqtgraph as pg
+from acq4.devices.Device import Device
+from acq4.util import Qt
+from acq4.util.future import future_wrap, Future, FutureButton
+
+
+class Sonicator(Device):
+    """Base class for any sonicator device."""
+
+    sigSonicationChanged = Qt.pyqtSignal(float)
+
+    def __init__(self, *args, **kwargs):
+        super().__init__(*args, **kwargs)
+        self.actionLock = RLock()
+
+    @future_wrap
+    def doProtocol(self, mode: str, _future: Future = None, **kwargs) -> None:
+        with self.actionLock:
+            if mode in {"clean", "expel"}:
+                duration = kwargs.get("duration", 5 if mode == "clean" else 1)
+                frequency = kwargs.get("frequency", 150e3)
+                _future.waitFor(self.sonicate(frequency, duration, lock=False))
+            elif mode == "quick cleanse":
+                start = kwargs.get("start", 140e3)
+                stop = kwargs.get("stop", 154e3)
+                step = kwargs.get("step", 1e3)
+                step_duration = kwargs.get("stepDuration", 100e-3)
+                frequency = start
+                while frequency <= stop:
+                    _future.waitFor(self.sonicate(frequency, step_duration, lock=False))
+                    frequency += step
+                frequency -= step
+                while frequency > start:
+                    _future.waitFor(self.sonicate(frequency, step_duration, lock=False))
+                    frequency -= step
+            else:
+                raise ValueError(f"Unrecognized sonication protocol '{mode}'")
+
+    def sonicate(self, frequency: float, duration: float, lock: bool = True) -> Future:
```

Comment:
Use device reservation instead of passing a lock around? And in the subclass, make sure to reserve both devices together.

Reply:
I think it is enough to just reserve the daq generic device (and the daq). I've removed other locks.